src/recluster.c

```
@@ -0,0 +1,1175 @@
+				  MultiXactId cutoffMulti);
+
+static void swap_relation_files(Oid r1, Oid r2,
+					bool swap_toast_by_content,
+					bool is_internal,
+					TransactionId frozenXid,
+					MultiXactId cutoffMulti);
+
+/*
+ * timescale_cluster_rel
+ *
+ * This clusters the table by creating a new, clustered table and
+ * swapping the relfilenodes of the new table and the old table, so
+ * the OID of the original table is preserved.  Thus we do not lose
+ * GRANT, inheritance nor references to this table (this was a bug
+ * in releases through 7.3).
+ *
+ * Indexes are rebuilt in the same manner.
+ *
+ * If indexOid is InvalidOid, the table will be rewritten in physical order
+ * instead of index order.
+ */
+void
+timescale_recluster_rel(Oid tableOid, Oid indexOid, bool recheck, bool verbose)
+{
+	Relation	OldHeap;
+
+	if (!OidIsValid(indexOid))
+		elog(ERROR, "Recluster must specify an index.");
+
+	/* Check for user-requested abort. */
+	CHECK_FOR_INTERRUPTS();
+
+	/*
+	 * We grab exclusive access to the target rel and index for the duration
+	 * of the transaction.  (This is redundant for the single-transaction
+	 * case, since cluster() already did it.)  The index lock is taken inside
+	 * check_index_is_clusterable.
+	 */
+	OldHeap = try_relation_open(tableOid, ExclusiveLock);
+
+	/* If the table has gone away, we can skip processing it */
+	if (!OldHeap)
+		return;
+
+	/*
+	 * Since we may open a new transaction for each relation, we have to check
+	 * that the relation still is what we think it is.
+	 *
+	 * If this is a single-transaction CLUSTER, we can skip these tests. We
+	 * *must* skip the one on indisclustered since it would reject an attempt
+	 * to cluster a not-previously-clustered index.
+	 */
+	if (recheck)
+	{
+		HeapTuple	tuple;
+		Form_pg_index indexForm;
+
+		/* Check that the user still owns the relation */
+		if (!pg_class_ownercheck(tableOid, GetUserId()))
+		{
+			relation_close(OldHeap, ExclusiveLock);
+			return;
+		}
+
+		/*
```

Let's throw an error for any kind of temp table. "Other" or not.